### PR TITLE
Diagnostics file writes use RedactSecretPaths

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

The elastic-agent will use redact secret paths in files written in diagnostics bundles. Secret paths are expected to be specified as a top-level attribute in yaml data being written.

## Why is it important?

Secrets that the fleet-server injects into policies can appear in diagnostics bundles.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

## Disruptive User Impact

Secrets in diagnostics bundles will be redacted, as our [public documentation discloses](https://www.elastic.co/guide/en/fleet/current/elastic-agent-cmd-options.html#elastic-agent-diagnostics-command).

## How to test this PR locally

Enrol in a fleet policy with secrets and collect a diagnostics bundle.

### CODE REVIEW

File: internal/pkg/diagnostics/diagnostics.go [hunk 1]
```diff
@@ -327,12 +327,13 @@ func writeRedacted(errOut, resultWriter io.Writer, fullFilePath string, fileResu
 
 	// Should we support json too?
 	if fileResult.ContentType == "application/yaml" {
-		unmarshalled := map[interface{}]interface{}{}
```
Comment (michel-laterman) [inline internal/pkg/diagnostics/diagnostics.go:330]: I think unmarshaling to `map[interface{}]interface{}` was a restriction with yaml.v2 which is not a part of yaml.v3, see https://github.com/go-yaml/yaml/issues/139#issuecomment-2247444123

File: internal/pkg/diagnostics/diagnostics.go [hunk 2]
```diff
@@ -579,7 +580,6 @@ func saveLogs(name string, logPath string, zw *zip.Writer) error {
 func RedactSecretPaths(mapStr map[string]any, errOut io.Writer) map[string]any {
 	v, ok := mapStr["secret_paths"]
 	if !ok {
-		fmt.Fprintln(errOut, "No output redaction: secret_paths attribute not found.")
```
Comment (michel-laterman) [inline internal/pkg/diagnostics/diagnostics.go:582]: When a diagnostics action is ran secrets in the component related files are not redacted (structure does not match, and `secret_paths` is not passed)
This line causes extra noise in the output

File: testing/fleetservertest/checkin.go [hunk 3]
```diff
@@ -174,11 +174,11 @@ const (
             "hosts": {{ toJson .FleetHosts }}
           },
           "id": "{{.PolicyID}}",
-          "secret_paths": ["inputs.0.secret_key"],
+          "secret_paths": ["inputs.0.custom_attr"],
           "inputs": [
             {
               "id": "fake-input",
-              "secret_key": "secretValue",
+              "custom_attr": "secretValue",
```
Comment (michel-laterman) [inline testing/fleetservertest/checkin.go:181]: I've renamed to `custom_attr` to ensurue `redactKey` does not match the attribute key

File: testing/integration/diagnostics_test.go [hunk 4]
```diff
@@ -20,11 +20,14 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v2"
```
Comment (blakerouse) [inline testing/integration/diagnostics_test.go:23]: Above you where talking about yaml.v3 but here you import yaml.v2? Which one is being used here? We should unify our yaml usage to a single library.

Comment (michel-laterman) [reply testing/integration/diagnostics_test.go]: I've updated this and all the integration test uses, and have made https://github.com/elastic/elastic-agent/issues/5750 to do the rest

**Review by blakerouse — APPROVED** [review]
Thanks for updating to yaml.v3. This looks good!

File: internal/pkg/diagnostics/diagnostics_test.go [hunk 5]
```diff
@@ -140,6 +140,103 @@ mapping:
...
+`),
+		expect: `id: test-policy
+inputs:
+    - redactKey: secretValue
+      type: test_input
+outputs:
+    default:
+        api_key: <REDACTED>
+        redactOtherKey: secretOutputValue
+        type: elasticsearch
```
Comment (swiatekm) [inline internal/pkg/diagnostics/diagnostics_test.go:168]: I think these would be more readable if inputs and outputs used the same indentation.

**Comment by jlind23** [comment]
The error message returned while trying to create the ESS stack is weird..
`error deploying stack for request 8160-SNAPSHOT: Non-201 status code returned by server: 504, body: {"errors": [{"message": "The service timed out", "code": "root.timeout"}]}`

Looks like the version digit separator '.' has been removed for `8.16.0-SNAPSHOST`, not sure if that's the actual issue though.

**Comment by jlind23** [comment]
Looks like tests are panicking, weirdly enough this seems to be related to the yaml.V3 bump.
```
--- FAIL: TestInstallWithBasePath (27.97s)
panic: interface conversion: interface {} is map[string]interface {}, not map[interface {}]interface {} [recovered]
	panic: interface conversion: interface {} is map[string]interface {}, not map[interface {}]interface {}

goroutine 84 [running]:
testing.tRunner.func1.2({0x4368f40, 0xc0006af6e0})
	/usr/local/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1634 +0x377
panic({0x4368f40?, 0xc0006af6e0?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
github.com/elastic/elastic-agent/testing/integration.unmarshalVersionOutput(0xc00052e000, {0xc000855400, 0x81, 0x400}, {0x480c968, 0x6})
	/home/ubuntu/agent/testing/integration/pkgversion_common_test.go:106 +0x1d8
github.com/elastic/elastic-agent/testing/integration.TestInstallWithBasePath.testAgentPackageVersion.func1(0xc00052e000)
	/home/ubuntu/agent/testing/integration/pkgversion_common_test.go:48 +0x27a
testing.tRunner(0xc00052e000, 0xc0006aef30)
	/usr/local/go/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 76
	/usr/local/go/src/testing/testing.go:1742 +0x390
```

**Comment by michel-laterman** [comment]
I'm going to roll back some of the updates to yaml.v3

**Comment by ycombinator** [comment]
Not sure why SonarQube is reporting 0% coverage on new code.  The unit test `TestRedactWithSecretPaths` was added and it exercises code that was changed in the `writeRedacted` function.   Force merging.

**Pull request merged**